src/classes/SRVC_ContactPrimaryLanguage.cls

```diff
@@ -0,0 +1,234 @@
+    *         be inserted and one list containing Contact Language records to be updated.                                         
+    ************************************************************************************************************/
+    @TestVisible
+    protected virtual ContactLanguagesToProcess getPrimaryContactLanguagesToInsertOrUpdate(Map<Id, Contact_Language__c> contactLanguagesToInsertByContactId,
+                                                                                    Map<Id, Id> languageIdToCheckForUpdateByContactId){
+
+        List<Contact_Language__c> contactLanguagesToUpdateToPrimaryList = new List<Contact_Language__c>();
+
+        List<Contact_Language__c> existingContactLanguagesList = [SELECT Id,
+                                                                          Language__c,
+                                                                          Contact__c,
+                                                                          Primary_Language__c
+                                                                  FROM Contact_Language__c
+                                                                  WHERE Contact__c IN :languageIdToCheckForUpdateByContactId.keySet()
+                                                                  AND (Language__c IN :languageIdToCheckForUpdateByContactId.values()
+                                                                                   OR Primary_Language__c = true)];
+                                                                        
+        for (Contact_Language__c conLang : existingContactLanguagesList){
+            if (languageIdToCheckForUpdateByContactId.containsKey(conLang.Contact__c) &&
+                languageIdToCheckForUpdateByContactId.get(conLang.Contact__c) == conLang.Language__c){
+                
+                conLang.Primary_Language__c = true;
+                contactLanguagesToUpdateToPrimaryList.add(conLang);
+                contactLanguagesToInsertByContactId.remove(conLang.Contact__c);
+            } 
+        }
+
+        ContactLanguagesToProcess conLangsToProcess = new ContactLanguagesToProcess(contactLanguagesToInsertByContactId.values(), 
+                                                                                            contactLanguagesToUpdateToPrimaryList);
+
+        return conLangsToProcess;
+    }
+
+    /************************************************************************************************************
+    * @description Value object specifying a list of new Contact Language records to be inserted 
+    * and a list of existing Contact Language records to be updated.
+    ************************************************************************************************************/
+    public class ContactLanguagesToProcess{
+        public List<Contact_Language__c> contactLanguagesToInsertList { get; private set; }
```

Comment:
@m-kinkade This is an extremely efficient way to handle the situation with the access controls we desired.